File: matlab/util/MZtest.m
function [MZstat MZpval]=MZtest(y,yhat)
   
  for ii=1:size(yhat,2)
    intercept=ones(length(y),1); X=[intercept yhat(:,ii)];
    [x,resnormU] = lsqlin(X,y); SSE=sum(resnormU);
    A=eye(2); b=[0;1]; N_restriction=length(b); [x,resnormR] = lsqlin(X,y,A,b); SSER=sum(resnormR);
    N=length(y); k=size(X,2); df=N-k; %degrees of freedom  
    MZstat(ii,1)=((SSER-SSE)/N_restriction)/(SSE/df);
    MZpval(ii,1)=1-fcdf(MZstat(ii,1),N_restriction,N-k);
  end
   
end


% %F-test
%   %Simulated Data
% T=1000;
% ResidualVar=0.1;
% beta=1;
% alfa=0;
% yhat=normrnd(0,1,T,1); y=alfa*ones(T,1)+yhat*beta+normrnd(0,ResidualVar,T,1);
%   %Real Data
% 
% scatter(yhat,y)
% intercept=ones(length(y),1); X=[intercept yhat];
% %Unrestricted
% [x,resnormU] = lsqlin(X,y); SSE=sum(resnormU);
% %Restricted
% A=eye(2); b=[0;1]; N_restriction=length(b); [x,resnormR,residual] = lsqlin(X,y,A,b); SSER=sum(resnormR);
% %Compute MZtest
% N=length(y); k=size(X,2); df=N-k; %degrees of freedom
% Ftest=((SSER-SSE)/N_restriction)/(SSE/df);
% pval=1-fcdf(Ftest,N_restriction,N-k);

File: python/mztest.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 21

@author: victorsellemi
"""

import numpy as np
import numpy.linalg as LA
from scipy import optimize     
from scipy.stats import f        

def MZtest(y,yhat):
    """
    DESCRIPTION: 
        function to implement Mincer-Zarnowitz Test
    INPUT:
        y = true series to be forecasted
        yhat = forecasted series
    OUTPUT:
        MZStat = MZ test statistic
        MZpval = MZ test p_value    
    """
    MZstat = []
    MZpval = []
    for ii in range(yhat.shape[1]):
        intercept = np.ones((y.shape[0],1))
        X = np.concatenate((intercept, yhat[:,[ii]]), axis = 1)
        x,_, _, _= LA.lstsq(X,y, rcond = None)
        resnormU = np.sum((x[0]*X[:,[0]] + x[1]*X[:,[1]] - y)**2)
        SSE = np.sum(resnormU)
        N_restriction = 2
        cons = ({'type' : 'ineq', 'fun': lambda x: -x[0]},
                {'type': 'ineq', 'fun': lambda x: -x[1] + 1} )
        sol = optimize.minimize(lambda x: np.sum((x[0]*X[:,[0]] + x[1]*X[:,[1]] - y)**2), x0 = np.ones((2,1)), constraints = cons)
        #sol = optimize.minimize(lambda x: LA.norm(X.dot(x) - y), x0 = np.array([[1],[1]]), constraints = cons)
        x = sol.x
        resnormR = np.sum((x[0]*X[:,[0]] + x[1]*X[:,[1]] - y)**2)
        SSER = np.sum(resnormR)
        N = len(y)
        k = X.shape[1]
        df = N - k
        temp_MZstat = ((SSER - SSE) / N_restriction) / (SSE / df) 
        MZstat.append(temp_MZstat)
        MZpval.append( 1 - f.cdf(temp_MZstat, N_restriction, df))
    return np.array(MZstat), np.array(MZpval)
